src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -244,6 +248,99 @@ public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions
             }
         }
 
+        internal override T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+#if NETCOREAPP
+            bool success = TryParseEnumCore(ref reader, options, out T value);
+#else
+            bool success = TryParseEnumCore(reader.GetString(), options, out T value);
+#endif
+
+            if (!success)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+            return value;
+        }
+
+        internal override void WriteAsPropertyNameCore(Utf8JsonWriter writer, T value, JsonSerializerOptions options, bool isWritingExtensionDataProperty)
+        {
+            ulong key = ConvertToUInt64(value);
```

Comment:
Given that you've now keying the caches on `T` rather than `ulong`, shouldn't this moved to the bottom of the method after we're done trying to serialize as string?

Reply:
Ah, hadn't realized you reverted that change. Nevermind.